core/src/main/scala/kafka/coordinator/transaction/TransactionManager.scala

```diff
@@ -0,0 +1,294 @@
+              }
+
+              currOffset = entry.nextOffset
+            }
+          }
+
+          loadedPids.foreach {
+            case (transactionalId, pidMetadata) =>
+              val currentPidMetadata = addPid(transactionalId, pidMetadata)
+              if (!pidMetadata.equals(currentPidMetadata)) {
+                // treat this as a fatal failure as this should never happen
+                fatal(s"Attempt to load $transactionalId's pid metadata $pidMetadata failed " +
+                  s"because there is already a different cached pid metadata $currentPidMetadata; " +
+                  s"all future client requests with transactional ids related to this partition will result in an error code, " +
+                  s"and this partition of the log will be effectively disabled.")
+
+                corruptedPartitions.add(topicPartition.partition)
+              }
+          }
+
+          removedPids.foreach { transactionalId =>
+            // if the cache already contains a pid which should be removed, raise an error.
+            if (pidMetadataCache.contains(transactionalId))
+              throw new IllegalStateException(s"Unexpected unload of $transactionalId's pid metadata while " +
+                s"loading partition $topicPartition")
+          }
+
+          info(s"Finished loading ${loadedPids.size} pid metadata from $topicPartition in ${time.milliseconds() - startMs} milliseconds")
+        }
+    }
+  }
+
+  /**
+    * When this broker becomes a leader for a transaction log partition, load this partition and
+    * populate the pid metadata cache with the transactional ids.
+    */
+  def loadTransactionsForPartition(partition: Int) {
+    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partition)
+
+    def loadPidAndTransactions() {
+      info(s"Loading pid metadata from $topicPartition")
+
+      inLock(partitionLock) {
+        if (loadingPartitions.contains(partition)) {
+          // with background scheduler containing one thread, this should never happen,
+          // but just in case we change it in the future.
+          info(s"Pid and transaction status loading from $topicPartition already in progress.")
+          return
+        } else {
+          loadingPartitions.add(partition)
+        }
+      }
+
+      try {
+        loadPidMetadata(topicPartition)
+      } catch {
+        case t: Throwable => error(s"Error loading offsets from $topicPartition", t)
+      } finally {
+        inLock(partitionLock) {
+          ownedPartitions.add(partition)
+          loadingPartitions.remove(partition)
+        }
+      }
+    }
+
+    scheduler.schedule(topicPartition.toString, loadPidAndTransactions)
+  }
+
+  /**
+    * When this broker becomes a follower for a transaction log partition, clear out the cache for corresponding transactional ids
+    * that belong to that partition.
+    */
+  def removeTransactionsForPartition(partition: Int) {
```

Comment:
Users can still change the partition counts in Zk directly (at least one of our customers has done this with the offsets topic). We don't need to do it here, but I was just thinking it might make the code a bit more resilient.